File: Основы программирования/Лабораторные работы/18/Pascal/U_SORT.pas
UNIT U_SORT;

{-----------------------------------------------------------------}
							INTERFACE
{-----------------------------------------------------------------}	

	type
		t_func = function(var a, b): boolean;
	
	procedure swap(var a, b; size: byte);
	procedure sort(var a; a_length, el_size: byte; func: t_func);


{-----------------------------------------------------------------}
							IMPLEMENTATION
{-----------------------------------------------------------------}

	procedure swap(var a, b; size: byte);

		var 
			p: pointer;
    
    	begin
        	
        	getmem(p, size);
        	move(a, p^, size);
        	move(b, a, size);
        	move(p^, b, size);
        	freemem(p, size);

		end;


	procedure sort(var arr; a_length, el_size: byte; func: t_func);

		type	
			pbyte = ^byte;

		var
			start, first, second, max: pbyte;
			i, j: byte;

		begin
			
			start := pbyte(arr);

			for i := 0 to a_length - 2 do
				begin
					first := pbyte(longword(start) + i*el_size);
					max := first;
					for j := i + 1 to a_length - 1 do
						begin
							second := pbyte(longword(start) + j*el_size);
							if (func(second^, max^)) then
								max := second;
						end;
					swap(first^, max^, el_size);
				end;

		end;

begin
	
	//nothing to do here

end.

File: Основы программирования/Лабораторные работы/18/C/u_sort.c
#include <stdio.h>
#include <stdlib.h>
#include "u_sort.h"

void swap(void *a, void *b, size_t size)
{
	char tmp;
	int i = 0;
	for (; i < size; i++) {
		tmp = ((char *)a)[i];
		((char *)a)[i] = ((char *)b)[i];
		((char *)b)[i] = tmp;
	}
}

void sort(void * arr, unsigned length, size_t size, int (*funct)(void *, void *))
{
	char *begin = (char *)arr, *max;
	int i = 0, j;
	for (;i < length - 1; i++) 
	{
		max = begin + i * size;
		for (j = i + 1; j < length; j++)
			if (funct(begin + j * size, max)) {
				max = begin + j * size;
			}
		swap(begin + i * size, max, size);
	}
}